src/librustc_typeck/collect.rs

```diff
@@ -2522,6 +2522,51 @@ fn codegen_fn_attrs(tcx: TyCtxt<'_>, id: DefId) -> CodegenFnAttrs {
                     }
                 }
             }
+        } else if tcx.sess.check_name(attr, sym::instruction_set) {
+            // CHECK tcx.sess.target.target supports switching
+            if tcx.sess.target.target.options.has_thumb_interworking {
+                codegen_fn_attrs.instruction_set = match attr.meta().map(|i| i.kind) {
+                    Some(MetaItemKind::List(ref items)) => {
+                        if items.len() != 2 {
+                            struct_span_err!(
+                                tcx.sess.diagnostic(),
+                                attr.span,
+                                E0772,
+                                "expected one argument"
+                                ).emit();
+                            None
+                        } else if items[0].has_name(sym::arm) {
+                            if items[1].has_name(sym::a32) {
+                                Some(InstructionSetAttr::ArmA32)
+                            } else if items[1].has_name(sym::t32) {
+                                Some(InstructionSetAttr::ArmT32)
+                            } else {
+                                None
+                            }
+                        } else {
+                            struct_span_err!(
+                                tcx.sess.diagnostic(),
+                                items[0].span(),
+                                E0773,
+                                "invalid argument"
+                            )
+                            .emit();
+                            None
+                        }
+                    }, 
+                    _ => {
+                        // this shouldn't happen I don't think?
+                        None
+                    }
+                };
+            } else {
+                struct_span_err!(
+                    tcx.sess.diagnostic(),
+                    attr.span,
+                    E0773,
+                    "invalid argument"
```

Comment:
A better message here would be nice, maybe something like "target x-y-z does not support `#[instruction_set]`"?